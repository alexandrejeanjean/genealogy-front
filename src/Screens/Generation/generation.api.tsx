import GenerationModel from "../../models/generation.model";
import { ErrorHandler } from "../../helpers";
import apiClient from "../../api";

/**
 * get list of generations for the given user
 * @returns {Array<Generations>} - return list of generation with peoples
 */
export async function apiGetGenerations(
  setToastVisible: Function,
  familyId: number
) {
  try {
    const response = await apiClient.get(`/families/${familyId}/generations`);
    const generations = response.data;
    if (generations) {
      return generations.map(
        (Generation: any) => new GenerationModel(Generation)
      );
    }
  } catch (err) {
    ErrorHandler(
      err,
      setToastVisible,
      "We can't access the generation's list."
    );
  }
}

/**
 * post new generation for the given user
 * @param {Function} setToastVisible - function to load modal
 * @param {Object} gen - generation's data
 * @param {Number} familyId - id of the family to create a new generation
 * @returns return 201 if new generation created successfully
 */

export async function apiCreateGeneration(
  setToastVisible: Function,
  gen: { position: number },
  familyId: number
) {
  try {
    const response = await apiClient.post(`/families/${familyId}/generations`, {
      position: gen.position,
    });

    if (response) {
      return response;
    }
  } catch (err) {
    ErrorHandler(err, setToastVisible, "We can't create a new generation.");
  }
}

/**
 * delete a generation for the given user by id
 * @param {Function} setToastVisible - function to load modal
 * @param {Number} familyId - id of the family
 * @param {Number} generationId - id of the generation to delete
 * @returns return 204 if generation deleted successfully
 */

export async function apiDeleteGeneration(
  setToastVisible: Function,
  familyId: number,
  generationId: number
) {
  try {
    const response = await apiClient.delete(
      `/families/${familyId}/generations/${generationId}`
    );

    if (response) return response.data;
  } catch (err) {
    ErrorHandler(err, setToastVisible, "We can't delete this generation.");
  }
}
