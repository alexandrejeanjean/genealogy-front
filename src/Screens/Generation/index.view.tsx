import React from "react";

import { Container, Button, Row, Col } from "react-bootstrap";

import ListWrapper from "../../SharedComponents/ListWrapper/ListWrapper";
import ModalForm from "../../SharedComponents/ModalForm/ModalForm";
import CreateItemBtn from "../../SharedComponents/CreateItemBtn/CreateItemBtn";

import { deletePic } from "../../assets/imgPath";

type TPerson = {
  id: number;
  familyId: number;
  generationId: number;
  firstname: string;
  lastname: string;
};

type TGenerationItem = {
  id: number;
  position: number;
  familyId: number;
  peoples: TPerson[];
};

type TGetGenerations = TGenerationItem[];

type TFamilyRole = { id: number; role: string };

type TFamilyRoles = TFamilyRole[];

type TGeneration = {
  genList: TGetGenerations;
  modalPersonShow: { isVisible: boolean; additionalDatas: number | null };
  setModalPerson: Function;
  familyRoles: TFamilyRoles;
  createPerson: Function;
  deletePerson: Function;
  deleteGeneration: Function;
  getAvatar: Function;
  getRoleName: Function;
};

const Generation = ({
  genList,
  modalPersonShow,
  familyRoles,
  setModalPerson,
  createPerson,
  deletePerson,
  deleteGeneration,
  getAvatar,
  getRoleName,
}: TGeneration) => {
  return (
    <>
      <ModalForm
        title="Person"
        show={modalPersonShow.isVisible}
        inputs={[
          { name: "firstname", placeholder: "Ex: Luis", inputType: "text" },
          {
            name: "lastname",
            placeholder: "Ex: Armstrong",
            inputType: "text",
          },
          {
            name: "role",
            placeholder: "Ex: Grand-father",
            inputType: "select",
            datas: familyRoles,
          },
        ]}
        onHide={() =>
          setModalPerson({ isVisible: false, additionalDatas: null })
        }
        submit={createPerson}
      />

      {genList &&
        genList.map((gen: TGenerationItem) => {
          return (
            <Container fluid key={gen.position}>
              <Row>
                <Col className="generations-wrapper">
                  <Row>
                    <Col className="generation-head">
                      <p>Génération : {gen.position}</p>
                      <button
                        className="delete-btn"
                        onClick={() => deleteGeneration(gen.id)}
                      >
                        <img src={deletePic} alt="delete button" />
                      </button>
                    </Col>
                  </Row>

                  <ListWrapper
                    datas={gen.peoples}
                    getAvatar={(roleId: number) => getAvatar(roleId)}
                    getSubText={(roleId: number) => getRoleName(roleId)}
                    handleClick={() => console.log("click")}
                    handleDelete={(person: TPerson) => deletePerson(person)}
                  >
                    <CreateItemBtn
                      handleClick={() =>
                        setModalPerson({
                          isVisible: true,
                          additionalDatas: gen.id,
                        })
                      }
                    />
                  </ListWrapper>
                </Col>
              </Row>
            </Container>
          );
        })}
    </>
  );
};

export default Generation;
