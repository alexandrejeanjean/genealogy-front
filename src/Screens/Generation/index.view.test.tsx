import React from "react";
import { shallow } from "enzyme";
import Generation from "./index.view";

const GenerationProps = {
  genList: [
    {
      id: 0,
      position: 1,
      familyId: 1,
      peoples: [
        {
          id: 0,
          familyId: 1,
          generationId: 1,
          firstname: "Dark",
          lastname: "Vador",
        },
      ],
    },
  ],
  modalPersonShow: { isVisible: false, additionalDatas: null },
  setModalPerson: jest.fn(),
  familyRoles: [{ id: 0, role: "father" }],
  createPerson: jest.fn(),
  deletePerson: jest.fn(),
  deleteGeneration: jest.fn(),
  getAvatar: jest.fn(),
  getRoleName: jest.fn(),
};

describe("Generation", () => {
  it("should render correctly", () => {
    const Component = shallow(<Generation {...GenerationProps} />);
    expect(Component).toMatchSnapshot();
  });
});
