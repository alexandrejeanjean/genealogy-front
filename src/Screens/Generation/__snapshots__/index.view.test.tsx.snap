// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generation should render correctly 1`] = `
<Fragment>
  <ModalForm
    inputs={
      Array [
        Object {
          "inputType": "text",
          "name": "firstname",
          "placeholder": "Ex: Luis",
        },
        Object {
          "inputType": "text",
          "name": "lastname",
          "placeholder": "Ex: Armstrong",
        },
        Object {
          "datas": Array [
            Object {
              "id": 0,
              "role": "father",
            },
          ],
          "inputType": "select",
          "name": "role",
          "placeholder": "Ex: Grand-father",
        },
      ]
    }
    onHide={[Function]}
    show={false}
    submit={[MockFunction]}
    title="Person"
  />
  <Container
    fluid={true}
    key="1"
  >
    <Row
      noGutters={false}
    >
      <Col
        className="generations-wrapper"
      >
        <Row
          noGutters={false}
        >
          <Col
            className="generation-head"
          >
            <p>
              Génération : 
              1
            </p>
            <button
              className="delete-btn"
              onClick={[Function]}
            >
              <img
                alt="delete button"
                src={[Function]}
              />
            </button>
          </Col>
        </Row>
        <ListWrapper
          datas={
            Array [
              Object {
                "familyId": 1,
                "firstname": "Dark",
                "generationId": 1,
                "id": 0,
                "lastname": "Vador",
              },
            ]
          }
          getAvatar={[Function]}
          getSubText={[Function]}
          handleClick={[Function]}
          handleDelete={[Function]}
        >
          <CreateItemBtn
            handleClick={[Function]}
          />
        </ListWrapper>
      </Col>
    </Row>
  </Container>
</Fragment>
`;
